### DISCUSSION TITLE
Snippets - short pieces of scripts

### DISCUSSION THREAD

**Discussion by jakubczajczynski** [nicolaasburgers/definition-library-releases#37] [Ideas]
This touches another topic, about User Objects but I wanted to separate this Idea.  https://github.com/nicolaasburgers/definition-library-releases/discussions/25

---

You might be interested in looking into Snippets from the Metahopper plugin which are saved as User Objects. Although, I experienced some issues with creating Snippets with Metahopper, (The Save Snipett... button seems stubborn and sometimes dialog shows after many attempts to bring it on) with the help in organizing things provided by the Definition Library I might delve into it more. 

To support Snippets saved as User Objects probably you don't need to do anything extra, but maybe this will bring your attention to this more general problem. Not everything can/should be put into Clusters. Having pre-made small (and quick to place) chunks of definitions that are repetitive is very helpful. Metahopper for years is no longer maintained by Andrew Heumann, if you are interested in providing support for the User Objects maybe you will find a time to implement something similar/better to create User Objects made out of multiple components (as Snippets do it). I mentioned earlier that Metahopper seems to have some problems, and this new solution shouldn't be dependent on the abandoned plugin anyway.

One way to simply introduce Snippets to your plugin is by allowing placing not clusterable definitions directly inside other definitions. I know that this seems to be contrary to what your plugin should do, but maybe if Snippet is grouped it can be also updatable similar to clusters? Anyway, I don't expect updating/reverting already placed Snippets.
Storing Snippets as User Objects has the advantage that they are very easily accessible, just as any other GH component. This is not true (at least yet) for definitions stored in the Definition Library.

**Comment by nicolaasburgers** [comment]
I think this is a very intriguing idea - I'll have a look at snippets in Metahopper.

**Comment by nicolaasburgers** [comment]
I am actually quite keen on this idea now @jakubczajczynski and am in the process of refactoring the code to allow this, assuming that I don't find any technical reasons that would block my implementation - so far in my investigation it seems like it's possible.  I'll essentially enable groups (GH_Group) objects to be published as snippets. 

I'm working on two features in parallel at the moment: snippets and support for Microsoft accounts rather than GitHub (i.e. support for SharePoint and OneDrive) so I'm not sure which will be released first but I'll keep this thread up-to-date with progress.

**Comment by nicolaasburgers** [comment]
While I finalise SharePoint & OneDrive support (look out for it in a release soon) I've been thinking of snippets and have mocked up two possible mechanisms.  Could you let me know your thoughts or ask clever questions so I can further develop the design?

I saw a couple of people say that it's hard to refactor GH logic within a group because when you select components using a selection rectangle, the group they're in also gets selected, which gets in the way.  Also, I thought that groups are so easily lost and if the group was a published snippet then losing the group would also lose the knowledge about the snippet.

So I thought having a dashed rectangle is an interesting option.

Now for a bit more detailed info about how each option would work for you as a user:

**OPTION 1: GROUPS** - using standard Grasshopper groups as the container for a snippet
![Snippet_Group_Option](https://github.com/user-attachments/assets/e7689650-203e-4e03-a39a-d023928a148f)
When you go to publish, it will list ALL groups in the GH file (like it does for clusters now) and you can choose which one you'd like to publish as a snippet.   This would mean you could publish multiple new snippets in one go, although the list of groups could be extremely long indeed.
After publishing, the group would have the head-up display (name, version etc) appear above it like it does for clusters & Hops.
You can add components to this group or remove components like you would any other group and it would seem a natural conceptual fit to have groups used for snippets.

**OPTION 2: SKETCH OUTLINES** - using a standard Grasshopper polyline on the canvas as the outline for a snippet
![Snippet_Sketch_Option](https://github.com/user-attachments/assets/2d6aaa9a-8478-4d07-84c4-b9130fb1af1c)
Despite this option using polylines added to the GH file using the Sketch tool, I would suggest that the user wouldn't need to draw outlines themselves - rather, they'd select multiple components and then use a new menu option to publish them as a snippet.  This mechanism does mean you can only publish new snippets one at a time.
After publishing, the plug-in will draw the dashed line and display the name & version head up display above the line.
Adding new components to this snippet would be as easy as positioning it inside the lines; DefinitionLibrary will recognise that as a change to the snippet - no need to explicitly add it to the snippet.  However, if there's not enough room for this, I did have an idea in mind to offer new menu items that would have the plug-in automatically redraw the dashed outline to enclose the selected component(s):
![Snippet_MenuItems_MockUp](https://github.com/user-attachments/assets/eb6f9537-1089-4c0e-b840-77725749549c)

When you've published snippets they'll appear in the Import/manage window like this:
(note this does include the changes for SharePoint & OneDrive that aren't released yet):
![ImportAsSnippet_MockUp](https://github.com/user-attachments/assets/5bfe9590-aba7-4c95-8185-752a01914073)

Any feedback or suggestions welcome!  Could you think of another mechanism other than these two?

**Reply by nicolaasburgers** [reply]
@jakubczajczynski

**Reply by jakubczajczynski** [reply]
Sorry I didn't speak up earlier, I think the choice of groups over outlines was the right one! Fantastic work. I know it's Beta, so some of the limitations may exist because of that, but I'll ask anyway.

1. It's not that big of a problem, but in my definitions the color of groups matters. Even though I changed the default color of groups to white, snippets imported from DefinitionLibrary change to the "factory" color in Grasshopper, which is purple. Would it be possible to keep the group colors the same as when publishing the snippet?

2. When updating / rolling back a snippet, the connection of components from this snippet to components outside of it is lost, is it possible for Grasshopper not to break this connection? Now you have to be very careful or have an extremely tidy definition to quickly reconnect components to those in the snippet without causing any damage.

**Reply by nicolaasburgers** [reply]
Yes it is indeed beta and one thing you'll notice is that the modified (*) askerisk is often shown for snippets when they haven't in fact changed, sometimes this happens even directly after importing it onto the canvas.  This is something I will still need to work on to rectify but thought it wasn't a showstopper for now.

Thanks for your view that the groups was the correct choice - that's good to hear.

In terms of your points:

1. Yes it possible to do this, and I can store it in the settings file in the same way that the column widths for the import window are stored - I can add this in an upcoming version

2. I thought about reconnecting the wires when updating/rolling back a snippet but decided it was too complex at this point to do it, and that other features were more important.

If you consider all the ways that two versions of a snippet could differ, the rules to decide which component output outside the snippet should connect to which (updated/rolled back) component input, and vice versa for outputs, could be quite complex:
  - What if two components in the incoming snippet have the same input name & type?
  - Should it prefer ones that are closest to the original X,Y position?  
  - Should it consider the type of the component which has the input?
... and so on

Currently, unlike for clusters, the IDs (instance GUIDs) of component inputs & outputs changes in the process of publishing and updating, to allow for multiple copies of the same snippet to be imported onto the same document on the canvas, among other things. So I could keep better links between previous IDs to enable: "a connection between this input ID and this output ID was there before, so re-instate it". This is also complex but probably feasible.

I am open to suggestions for a set of rules I could/should implement - even if they cover only a few types of scenarios.
